scRNA-seq-advanced/02-celltype_assignment.Rmd

```diff
@@ -0,0 +1,465 @@
+
+
+## Cell type annotation with `SingleR`
+
+Get some cell data
+```{r}
+# Don't ask about creating the cache directories, just create them
+ExperimentHub::setExperimentHubOption("ASK", FALSE)
+AnnotationHub::setAnnotationHubOption("ASK", FALSE)
+
+# Get Data from the Blueprint & Encode projects from celldex
+blueprint_encode <- celldex::BlueprintEncodeData(ensembl = TRUE)
+```
+What is this `blueprint_encode`?
+
+```{r}
+blueprint_encode
+```
+
+A SummarizedExperiment is very similar to a SingleCellExperiment, except rather than having each column as a cell, each column is a sample.
+Otherwise, the components are very similar: each row is still a gene, for example, and additional data about the samples
+are stored in the `colData`.
+
+What information do we have for the samples?
+
+```{r}
+colData(blueprint_encode)
+```
+
+Let's see how many samples of each type we have:
+
+```{r}
+data.frame(colData(blueprint_encode)) %>% 
+  dplyr::count(label.main, label.fine)
+```
+
+### What does `SingleR` do?
+
+First we build a model that we can use for classifying expression.
+Some options we might explore: how to find choose genes to include in the model.
+`genes = "de"` is the default, using genes differentially expressed among the labels
+`genes = "sd"`uses genes with the most variability across the data set.
+
+This time we will do the training and classification in a single step, using the convenience function `SingleR::SingleR()`.
+
+```{r}
+singler_result <- SingleR::SingleR(
+  sce,
+  ref = blueprint_encode,
+  labels = blueprint_encode$label.main
+)
+```
+
+Plot the results:
+```{r}
+SingleR::plotScoreHeatmap(singler_result)
+```
+
+
+```{r}
+sce$celltype_main <- singler_result$pruned.labels
+```
+
+```{r}
+scater::plotUMAP(sce, colour_by = "celltype_main")
+```
+
+```{r}
+table(singler_result$pruned.labels)
+```
+
+We can do some cleanup using `forcats`.
+
+```{r}
+# combine all of the uncommon cell types
+common_labels <- singler_result$pruned.labels %>%
+  forcats::fct_lump_min(100)
+sce$celltype_common <- common_labels
+```
+
+
+```{r}
+scater::plotUMAP(sce, colour_by = "celltype_common")
+```
+
+We can also use the finer-grained cell type data.
+
+If we were classifying multiple samples, we would want to do the training only once!
+
+So let's separate the training from the fitting.
+First fit the model.
+
+```{r}
+singler_finemodel <- SingleR::trainSingleR(
+  blueprint_encode,
+  labels = blueprint_encode$label.fine,
+  # use DE to select genes (default)
+  genes = "de", 
+  # only use genes in the sce object
+  restrict = rownames(sce),
+  BPPARAM = BiocParallel::MulticoreParam(4)
+)
+```
+
+Now classify the cells based on that model.
+
+```{r}
+singler_result_fine <- SingleR::classifySingleR(
+  sce,
+  singler_finemodel,
+  BPPARAM = BiocParallel::MulticoreParam(4)
+)
+```
+
+
+What labels were assigned, and how many?
+
+```{r}
+table(singler_result_fine$labels)
```

Comment:
I think we want to print this sorted by count somehow, rather than alphabetical.